statsmodels/gam/gam.py

```diff
@@ -0,0 +1,767 @@
+                   scale=None, cov_type='nonrobust', cov_kwds=None, use_t=None,
+                   weights=None):
+        """fit model with penalized reweighted least squares
+
+        """
+
+        # alpha = alpha * len(y) * self.scale / 100
+        # TODO: we need to rescale alpha
+        endog = self.endog
+        k_exog_linear = self.k_exog_linear
+        wlsexog = self.exog  # smoother.basis
+        spl_s = self.penal.penalty_matrix(alpha=alpha)
+
+        nobs, n_columns = wlsexog.shape
+
+        # TODO what are these values?
+        if weights is None:
+            self.data_weights = np.array([1.] * nobs)
+        else:
+            self.data_weights = weights
+
+        if not hasattr(self, '_offset_exposure'):
+            self._offset_exposure = 0
+
+        self.scaletype = scale
+        # TODO: check default scale types
+        # self.scaletype = 'dev'
+        # during iteration
+        self.scale = 1
+
+        if start_params is None:
+            mu = self.family.starting_mu(endog)
+            lin_pred = self.family.predict(mu)
+        else:
+            lin_pred = np.dot(wlsexog, start_params) + self._offset_exposure
+            mu = self.family.fitted(lin_pred)
+        dev = self.family.deviance(endog, mu)
+
+        history = dict(params=[None, start_params], deviance=[np.inf, dev])
+        converged = False
+        criterion = history['deviance']
+        # This special case is used to get the likelihood for a specific
+        # params vector.
+        if maxiter == 0:
+            mu = self.family.fitted(lin_pred)
+            self.scale = self.estimate_scale(mu)
+            wls_results = lm.RegressionResults(self, start_params, None)
+            iteration = 0
+
+        for iteration in range(maxiter):
+
+            # TODO: is this equivalent to point 1 of page 136:
+            # w = 1 / (V(mu) * g'(mu))  ?
+            self.weights = self.data_weights * self.family.weights(mu)
+
+            # TODO: is this equivalent to point 1 of page 136:
+            # z = g(mu)(y - mu) + X beta  ?
+            wlsendog = (lin_pred + self.family.link.deriv(mu) * (endog - mu)
+                        - self._offset_exposure)
+
+            # this defines the augmented matrix point 2a on page 136
+            wls_results = penalized_wls(wlsendog, wlsexog, spl_s, self.weights)
+            lin_pred = np.dot(wlsexog, wls_results.params).ravel()
+            lin_pred += self._offset_exposure
+            mu = self.family.fitted(lin_pred)
+
+            # self.scale = self.estimate_scale(mu)
+            history = self._update_history(wls_results, mu, history)
+
+            if endog.squeeze().ndim == 1 and np.allclose(mu - endog, 0):
+                msg = "Perfect separation detected, results not available"
+                raise PerfectSeparationError(msg)
+
+            # TODO need atol, rtol
+            converged = _check_convergence(criterion, iteration, tol, 0)
```

Comment:
pass kwargname=0 here?  Since my knowledge of _check_convergence is "it has an informative name so I can guess what it does, but I don't know its signature off the top of my head", this is the only thing unclear about this line.